File: lib/java/openlinktoken/src/test/java/org/openlinktoken/attributes/FieldRegistryTest.java
/* SPDX-License-Identifier: MIT */
package org.openlinktoken.attributes;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertTrue;

import org.junit.jupiter.api.Test;

import org.openlinktoken.attributes.general.StringAttribute;
import org.openlinktoken.attributes.person.LastNameAttribute;

/** Tests default field registration and custom field lookup. */
class FieldRegistryTest {

    /** Verifies the default registry contains the built-in field identifiers. */
    @Test
    void testCreateDefaultLoadsBuiltInAttributes() {
        var registry = FieldRegistry.createDefault();
        assertTrue(registry.size() > 0);
        assertTrue(registry.getFieldIds().contains("FirstName"));
        assertTrue(registry.getFieldIds().contains("LastName"));
        assertTrue(registry.getFieldIds().contains("String"));
    }

    /** Verifies lookup returns the registered attribute instance. */
    @Test
    void testGetAttributeReturnsRegisteredInstance() {
        var registry = FieldRegistry.createDefault();
        var attribute = registry.getAttribute("FirstName");
        assertTrue(attribute.isPresent());
        assertEquals("FirstName", attribute.get().getName());
    }

    /** Verifies an unknown field identifier has no registered attribute. */
    @Test
    void testGetAttributeReturnsEmptyForUnknownField() {
        var registry = FieldRegistry.createDefault();
        var attribute = registry.getAttribute("NonExistent");
        assertFalse(attribute.isPresent());
    }

    /** Verifies field lookup returns its identifier and attribute class. */
    @Test
    void testGetFieldReturnsAttributeField() {
        var registry = FieldRegistry.createDefault();
        var field = registry.getField("LastName");
        assertTrue(field.isPresent());
        assertEquals("LastName", field.get().getFieldId());
        assertEquals(LastNameAttribute.class, field.get().getAttributeClass());
    }

    /** Verifies the builder can register multiple custom fields for one attribute instance. */
    @Test
    void testBuilderRegistersCustomField() {
        var attribute = new StringAttribute();
        var registry = FieldRegistry.Builder.fromDefaults()
                .register("MotherLastName", StringAttribute.class, attribute)
                .register("FatherLastName", StringAttribute.class, attribute)
                .build();

        assertTrue(registry.getFieldIds().contains("MotherLastName"));
        assertTrue(registry.getFieldIds().contains("FatherLastName"));

        var motherAttr = registry.getAttribute("MotherLastName");
        var fatherAttr = registry.getAttribute("FatherLastName");
        assertTrue(motherAttr.isPresent());
        assertTrue(fatherAttr.isPresent());
        assertEquals(attribute, motherAttr.get());
        assertEquals(attribute, fatherAttr.get());
    }

    /** Verifies building from defaults preserves the built-in fields. */
    @Test
    void testBuilderFromDefaultsIncludesBuiltIns() {
        var registry = FieldRegistry.Builder.fromDefaults().build();
        assertTrue(registry.getFieldIds().contains("FirstName"));
        assertTrue(registry.getFieldIds().contains("LastName"));
    }

    /** Verifies multiple field identifiers can share one attribute type and instance. */
    @Test
    void testMultipleFieldsSameAttributeType() {
        var stringAttr = new StringAttribute();
        var registry = new FieldRegistry.Builder()
                .register("Field1", StringAttribute.class, stringAttr)
                .register("Field2", StringAttribute.class, stringAttr)
                .register("Field3", StringAttribute.class, stringAttr)
                .build();

        assertEquals(3, registry.size());

        var field1 = registry.getField("Field1");
        var field2 = registry.getField("Field2");
        var field3 = registry.getField("Field3");

        assertTrue(field1.isPresent());
        assertTrue(field2.isPresent());
        assertTrue(field3.isPresent());

        assertNotNull(registry.getAttribute("Field1").orElse(null));
        assertNotNull(registry.getAttribute("Field2").orElse(null));
        assertNotNull(registry.getAttribute("Field3").orElse(null));
    }
}

File: lib/python/openlinktoken/src/test/openlinktoken/attributes/field_registry_test.py
# SPDX-License-Identifier: MIT


from openlinktoken.attributes.field_registry import FieldRegistry
from openlinktoken.attributes.general.string_attribute import StringAttribute
from openlinktoken.attributes.person.last_name_attribute import LastNameAttribute


class TestFieldRegistry:
    """Test cases for FieldRegistry."""

    def test_create_default_loads_built_in_attributes(self):
        """Test that create_default populates the registry with built-in attributes."""
        registry = FieldRegistry.create_default()
        assert registry.size() > 0
        assert "FirstName" in registry.get_field_ids()
        assert "LastName" in registry.get_field_ids()
        assert "String" in registry.get_field_ids()

    def test_get_attribute_returns_registered_instance(self):
        """Test resolving an attribute by field ID."""
        registry = FieldRegistry.create_default()
        attribute = registry.get_attribute("FirstName")
        assert attribute is not None
        assert attribute.get_name() == "FirstName"

    def test_get_attribute_returns_none_for_unknown_field(self):
        """Test that unknown field IDs return None."""
        registry = FieldRegistry.create_default()
        attribute = registry.get_attribute("NonExistent")
        assert attribute is None

    def test_get_field_returns_attribute_field(self):
        """Test resolving an AttributeField by field ID."""
        registry = FieldRegistry.create_default()
        field = registry.get_field("LastName")
        assert field is not None
        assert field.field_id == "LastName"
        assert field.attribute_class == LastNameAttribute

    def test_builder_registers_custom_field(self):
        """Test that the builder can register custom fields with shared attribute types."""
        attribute = StringAttribute()
        registry = (
            FieldRegistry.Builder.from_defaults()
            .register("MotherLastName", StringAttribute, attribute)
            .register("FatherLastName", StringAttribute, attribute)
            .build()
        )

        assert "MotherLastName" in registry.get_field_ids()
        assert "FatherLastName" in registry.get_field_ids()

        mother_attr = registry.get_attribute("MotherLastName")
        father_attr = registry.get_attribute("FatherLastName")
        assert mother_attr is attribute
        assert father_attr is attribute

    def test_builder_from_defaults_includes_built_ins(self):
        """Test that from_defaults pre-populates built-in attributes."""
        registry = FieldRegistry.Builder.from_defaults().build()
        assert "FirstName" in registry.get_field_ids()
        assert "LastName" in registry.get_field_ids()

    def test_multiple_fields_same_attribute_type(self):
        """Test that multiple fields can share the same attribute type."""
        string_attr = StringAttribute()
        registry = (
            FieldRegistry.Builder()
            .register("Field1", StringAttribute, string_attr)
            .register("Field2", StringAttribute, string_attr)
            .register("Field3", StringAttribute, string_attr)
            .build()
        )

        assert registry.size() == 3
        assert registry.get_field("Field1") is not None
        assert registry.get_field("Field2") is not None
        assert registry.get_field("Field3") is not None
        assert registry.get_attribute("Field1") is not None
        assert registry.get_attribute("Field2") is not None
        assert registry.get_attribute("Field3") is not None
